File: node-mcp-demo/src/utils/apiClient.ts
import axios, { AxiosInstance } from 'axios';

const API_BASE_URL = process.env.API_BASE_URL || 'https://as.imds.ai/apiStats/api';

export interface ApiKeyInfo {
  name: string;
  account: string;
  apiKey: string;
}

export interface ModelStats {
  requests: number;
  allTokens: number;
  inputTokens: number;
  costs: {
    total: number;
  };
}

export interface AggregatedStats {
  requests: number;
  allTokens: number;
  totalCost: number;
  inputTokens: number;
}

export interface KeyStatsResult extends ApiKeyInfo {
  stats: AggregatedStats;
  success: boolean;
  error?: string;
}

/**
 * 获取apiId
 */
export async function getApiId(apiKey: string): Promise<string> {
  try {
    const response = await axios.post(`${API_BASE_URL}/get-key-id`, { apiKey });
    if (response.data && response.data.success) {
      const apiId = response.data.data.id || response.data.data;
      return apiId;
    }
    throw new Error(`获取apiId失败: ${JSON.stringify(response.data)}`);
  } catch (error: any) {
    throw new Error(`获取apiId失败 (${apiKey}): ${error.message}`);
  }
}

/**
 * 获取统计数据
 */
export async function fetchStats(apiId: string, period: 'daily' | 'monthly'): Promise<any> {
  try {
    const response = await axios.post(`${API_BASE_URL}/user-model-stats`, {
      apiId,
      period
    });

    if (response.data && response.data.success) {
      return response.data;
    }
    throw new Error(`获取统计数据失败: ${JSON.stringify(response.data)}`);
  } catch (error: any) {
    const errorMsg = error.response?.data
      ? JSON.stringify(error.response.data)
      : error.message;
    throw new Error(`获取统计数据失败 (apiId: ${apiId}, period: ${period}): ${errorMsg}`);
  }
}

/**
 * 汇总所有模型的数据
 */
export function aggregateData(modelData: ModelStats[]): AggregatedStats {
  if (!modelData || modelData.length === 0) {
    return {
      requests: 0,
      allTokens: 0,
      totalCost: 0,
      inputTokens: 0
    };
  }

  const aggregated = modelData.reduce((acc, model) => {
    acc.requests += model.requests || 0;
    acc.allTokens += model.allTokens || 0;
    acc.totalCost += model.costs?.total || 0;
    acc.inputTokens += model.inputTokens || 0;
    return acc;
  }, {
    requests: 0,
    allTokens: 0,
    totalCost: 0,
    inputTokens: 0
  });

  return aggregated;
}

/**
 * 获取单个Key的统计数据（含重试机制）
 */
export async function getKeyStats(
  keyInfo: ApiKeyInfo,
  period: 'daily' | 'monthly',
  retries: number = 3
): Promise<KeyStatsResult> {
  let lastError: Error | undefined;

  for (let i = 0; i < retries; i++) {
    try {
      // 步骤1：获取apiId
      const apiId = await getApiId(keyInfo.apiKey);

      // 步骤2：获取统计数据
      const stats = await fetchStats(apiId, period);

      // 步骤3：汇总数据
      const aggregated = aggregateData(stats.data);

      return {
        ...keyInfo,
        stats: aggregated,
        success: true
      };
    } catch (error: any) {
      lastError = error;
      console.error(`尝试 ${i + 1}/${retries} 失败:`, error.message);

      // 如果不是最后一次重试，等待一秒后重试
      if (i < retries - 1) {
        await new Promise(resolve => setTimeout(resolve, 1000));
      }
    }
  }

  // 所有重试都失败
  console.error(`获取 ${keyInfo.name} (${keyInfo.account}) 的统计数据失败:`, lastError?.message);
  return {
    ...keyInfo,
    stats: {
      requests: 0,
      allTokens: 0,
      totalCost: 0,
      inputTokens: 0
    },
    success: false,
    error: lastError?.message
  };
}

/**
 * 批量获取所有Key的统计数据
 */
export async function getAllKeyStats(
  apiKeys: ApiKeyInfo[],
  period: 'daily' | 'monthly'
): Promise<KeyStatsResult[]> {
  console.log(`开始获取所有Key的${period === 'daily' ? '今日' : '本月'}统计数据...`);

  // 并发请求所有Key的数据
  const promises = apiKeys.map(keyInfo => getKeyStats(keyInfo, period));
  const results = await Promise.all(promises);

  // 统计成功和失败的数量
  const successCount = results.filter(r => r.success).length;
  const failCount = results.filter(r => !r.success).length;

  console.log(`统计完成: ${successCount} 成功, ${failCount} 失败`);

  return results;
}

File: python-mcp-demo/utils/api_client.py
import httpx
import os
from typing import List, Dict, Any, Optional
from dataclasses import dataclass

API_BASE_URL = os.getenv('API_BASE_URL', 'https://as.imds.ai/apiStats/api')


@dataclass
class ApiKeyInfo:
    """API Key信息"""
    name: str
    account: str
    apiKey: str


@dataclass
class AggregatedStats:
    """汇总统计数据"""
    requests: int = 0
    allTokens: int = 0
    totalCost: float = 0.0
    inputTokens: int = 0


@dataclass
class KeyStatsResult:
    """单个Key的统计结果"""
    name: str
    account: str
    apiKey: str
    stats: AggregatedStats
    success: bool
    error: Optional[str] = None


async def get_api_id(api_key: str) -> str:
    """获取apiId"""
    try:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{API_BASE_URL}/get-key-id",
                json={"apiKey": api_key},
                timeout=10.0
            )
            data = response.json()
            
            if data and data.get('success'):
                api_id = data['data'].get('id') or data['data']
                return api_id
            
            raise Exception(f"获取apiId失败: {data}")
    except Exception as e:
        raise Exception(f"获取apiId失败 ({api_key}): {str(e)}")


async def fetch_stats(api_id: str, period: str) -> Dict[str, Any]:
    """获取统计数据"""
    try:
        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{API_BASE_URL}/user-model-stats",
                json={"apiId": api_id, "period": period},
                timeout=10.0
            )
            data = response.json()
            
            if data and data.get('success'):
                return data
            
            raise Exception(f"获取统计数据失败: {data}")
    except Exception as e:
        raise Exception(f"获取统计数据失败 (apiId: {api_id}, period: {period}): {str(e)}")


def aggregate_data(model_data: List[Dict[str, Any]]) -> AggregatedStats:
    """汇总所有模型的数据"""
    if not model_data:
        return AggregatedStats()
    
    aggregated = AggregatedStats()
    
    for model in model_data:
        aggregated.requests += model.get('requests', 0)
        aggregated.allTokens += model.get('allTokens', 0)
        aggregated.totalCost += model.get('costs', {}).get('total', 0)
        aggregated.inputTokens += model.get('inputTokens', 0)
    
    return aggregated


async def get_key_stats(
    key_info: ApiKeyInfo,
    period: str,
    retries: int = 3
) -> KeyStatsResult:
    """获取单个Key的统计数据（含重试机制）"""
    last_error = None
    
    for i in range(retries):
        try:
            # 步骤1：获取apiId
            api_id = await get_api_id(key_info.apiKey)
            
            # 步骤2：获取统计数据
            stats = await fetch_stats(api_id, period)
            
            # 步骤3：汇总数据
            aggregated = aggregate_data(stats.get('data', []))
            
            return KeyStatsResult(
                name=key_info.name,
                account=key_info.account,
                apiKey=key_info.apiKey,
                stats=aggregated,
                success=True
            )
        except Exception as e:
            last_error = e
            print(f"尝试 {i + 1}/{retries} 失败: {str(e)}")
            
            # 如果不是最后一次重试，等待1秒后重试
            if i < retries - 1:
                import asyncio
                await asyncio.sleep(1)
    
    # 所有重试都失败
    print(f"获取 {key_info.name} ({key_info.account}) 的统计数据失败: {str(last_error)}")
    return KeyStatsResult(
        name=key_info.name,
        account=key_info.account,
        apiKey=key_info.apiKey,
        stats=AggregatedStats(),
        success=False,
        error=str(last_error)
    )


async def get_all_key_stats(
    api_keys: List[ApiKeyInfo],
    period: str
) -> List[KeyStatsResult]:
    """批量获取所有Key的统计数据"""
    import asyncio
    
    print(f"开始获取所有Key的{'今日' if period == 'daily' else '本月'}统计数据...")
    
    # 并发请求所有Key的数据
    tasks = [get_key_stats(key_info, period) for key_info in api_keys]
    results = await asyncio.gather(*tasks)
    
    # 统计成功和失败的数量
    success_count = sum(1 for r in results if r.success)
    fail_count = sum(1 for r in results if not r.success)
    
    print(f"统计完成: {success_count} 成功, {fail_count} 失败")
    
    return list(results)

